File: dynamic-programming/PartitionEqualSubsetSum.java
import java.util.Arrays;

/**
 * Problem:
 * Given a non-empty array nums containing only positive integers, find if the array can be
 * partitioned into two subsets such that the sum of elements in both subsets is equal.
 */
public class PartitionEqualSubsetSum {

    // Approach 1: Memoization (Top-Down)
    // Time Complexity: O(n * sum), where n is the number of elements and sum is the total sum.
    // Space Complexity: O(n * sum) for the memoization table + O(n) for recursion stack.
    public boolean canPartitionMemo(int[] nums) {
        int totalSum = 0;
        for (int num : nums) {
            totalSum += num;
        }

        if (totalSum % 2 != 0) {
            return false;
        }

        int target = totalSum / 2;
        Boolean[][] memo = new Boolean[nums.length][target + 1];
        return solve(nums, 0, target, memo);
    }

    private boolean solve(int[] nums, int index, int target, Boolean[][] memo) {
        if (target == 0) {
            return true;
        }
        if (index >= nums.length || target < 0) {
            return false;
        }
        if (memo[index][target] != null) {
            return memo[index][target];
        }

        // Include nums[index] or not
        boolean result = solve(nums, index + 1, target - nums[index], memo) || solve(nums, index + 1, target, memo);
        memo[index][target] = result;
        return result;
    }

    // Approach 2: Tabulation (Bottom-Up)
    // Time Complexity: O(n * sum)
    // Space Complexity: O(sum) for the DP table.
    public boolean canPartitionTab(int[] nums) {
        int totalSum = 0;
        for (int num : nums) {
            totalSum += num;
        }

        if (totalSum % 2 != 0) {
            return false;
        }

        int target = totalSum / 2;
        boolean[] dp = new boolean[target + 1];
        dp[0] = true;

        for (int num : nums) {
            for (int i = target; i >= num; i--) {
                dp[i] = dp[i] || dp[i - num];
            }
        }
        return dp[target];
    }

    public static void main(String[] args) {
        PartitionEqualSubsetSum partitionSolver = new PartitionEqualSubsetSum();
        int[] nums1 = {1, 5, 11, 5};

        System.out.println("Can partition " + Arrays.toString(nums1) + "? (Memoization): " + partitionSolver.canPartitionMemo(nums1));
        System.out.println("Can partition " + Arrays.toString(nums1) + "? (Tabulation): " + partitionSolver.canPartitionTab(nums1));

        int[] nums2 = {1, 2, 3, 5};

        System.out.println("Can partition " + Arrays.toString(nums2) + "? (Memoization): " + partitionSolver.canPartitionMemo(nums2));
        System.out.println("Can partition " + Arrays.toString(nums2) + "? (Tabulation): " + partitionSolver.canPartitionTab(nums2));
    }
}

File: dynamic-programming/partition_equal_subset_sum.py
# Partition Equal Subset Sum - Dynamic Programming

"""
Problem:
Given a non-empty array nums containing only positive integers, find if the array can be
partitioned into two subsets such that the sum of elements in both subsets is equal.
"""

class PartitionEqualSubsetSum:
    # Approach 1: Memoization (Top-Down)
    # Time Complexity: O(n * sum), where n is the number of elements and sum is the total sum.
    # Space Complexity: O(n * sum) for the memoization table + O(n) for recursion stack.
    def canPartitionMemo(self, nums: list[int]) -> bool:
        total_sum = sum(nums)
        if total_sum % 2 != 0:
            return False
        
        target = total_sum // 2
        memo = {}

        def solve(index, current_sum):
            if (index, current_sum) in memo:
                return memo[(index, current_sum)]
            if current_sum == target:
                return True
            if current_sum > target or index >= len(nums):
                return False

            # Include nums[index] or not
            result = solve(index + 1, current_sum + nums[index]) or solve(index + 1, current_sum)
            memo[(index, current_sum)] = result
            return result

        return solve(0, 0)

    # Approach 2: Tabulation (Bottom-Up)
    # Time Complexity: O(n * sum)
    # Space Complexity: O(sum) for the DP table.
    def canPartitionTab(self, nums: list[int]) -> bool:
        total_sum = sum(nums)
        if total_sum % 2 != 0:
            return False
        
        target = total_sum // 2
        dp = [False] * (target + 1)
        dp[0] = True

        for num in nums:
            for i in range(target, num - 1, -1):
                dp[i] = dp[i] or dp[i - num]
        
        return dp[target]

# Test cases
if __name__ == "__main__":
    partition_solver = PartitionEqualSubsetSum()
    nums1 = [1, 5, 11, 5]
    
    print(f"Can partition {nums1}? (Memoization): {partition_solver.canPartitionMemo(nums1)}")
    print(f"Can partition {nums1}? (Tabulation): {partition_solver.canPartitionTab(nums1)}")

    nums2 = [1, 2, 3, 5]
    
    print(f"Can partition {nums2}? (Memoization): {partition_solver.canPartitionMemo(nums2)}")
    print(f"Can partition {nums2}? (Tabulation): {partition_solver.canPartitionTab(nums2)}")
